README.md
# jQuery Modulider
Copyright &copy; [José Almeida](http://joseafga.com.br) <br />
License [Creative Commons Attribution 3.0 Unported License](http://creativecommons.org/licenses/by/3.0/ "License CC by 3.0")

Easy, Fast, Extensible and Flexible Slider for jQuery

## Demonstration:
[jQuery Modulider Demo](http://janightmare.github.io/jquery.modulider/demo/)

## Tested and Working:
* Internet Explorer 6, 7, 8, 9, 10
* Firefox 3.6.8, 8.0, 9.0
* Opera 11.52
* Chrome 15.0
* Safari 5.0

## Features:
* Autoplay
* Show Arbitrary/Next/Previous Slide
* Captions
* Navigation
* Automatically Generate Thumbnails for Navigation
* Preload
* Custom Transitions
* Random Transition
* Random Transition List - *you choose the transitions to random*
* Easy Customization
* Modular
* Events

## Options:
* `slides` jQuery Object
* `captions` jQuery Object
* `nav` jQuery Object
* `delay` Integer
* `speed` Integer
* `transition` String
* `randomList` Array
* `autoNav` Boolean
* `preload` Boolean
* `autoplay` Boolean
* `mouseOver` Boolean

## Main Methods:
* `.play()`
* `.pause()`
* `.next()`
* `.prev()`
* `.goTo(Integer to)`

## Default Transitions:
* none
* random
* fade
* crossfade
* fadebackground
* slide
* crossslide
* slidebackground
* show
* crossshow
* showbackground

## Events:
* onPlay
* onPause
* onNext
* onPrev
* onStartChanging `vars: current slide(int), next slide(int)`
* onFinishChanging

CHANGELOG.md
CHANGELOG:
===================

Version 0.9.1 (2013-02-01)
--------------------
* First public development version

demo/js/dev.js
$(function (){ // document ready
	var devSlider = window.slider;
	/* begin monitor */
	// round with precision
	Math._round = function (x, precision){
		var p = Math.pow(10, precision || 1);
		return Math.round(x * p) / p;
	};

	setInterval(function (){
		$('#current').text(devSlider.currSlide);
		$('#next').text(devSlider.nextSlide);
		$('#running').text(devSlider.inTransition);
	}, 10);

	// select transition
	var transitions = '<option value="select">select...</option>';
	$.each($.fn.modulider.transitions, function (key){
		transitions += '<option value="' + key + '">' + key + '</option>';
	});

	$('#transitions').change(function (){
		if (this.value != 'select')
			devSlider.options.transition = this.value;
	}).html(transitions);

	// mouseOver option
	$(':checkbox').each(function (){
		this.checked = false;
	});

	$('#mOver').click(function (e){
		if (this.checked) {
			devSlider.options.setSlideEvents.call(devSlider);
			devSlider.options.mouseOver = true;
		} else {
			devSlider.options.slides.off();
			devSlider.options.mouseOver = false;
		}
	});
	if (devSlider.options.mouseOver)
		$('#mOver').click();

	// delay and speed option
	$('#delay, #speed').each(function (){
		this.value = devSlider.options[this.id];
	}).change(function (){
		devSlider.options[this.id] = parseInt(this.value);
	});
	/* end monitor */

	/* begin controls */
	$('.play, .pause, .prev, .next').click(function (){
		devSlider[$(this).attr('class')]();
	});
	/* end controls */
});

src/jquery.modulider-plugins.js
(function ($) {
	/*!
	 * jQuery Modulider - Debug plugin
	 *
	 * Version 0.2.3
	 * Copyright José Almeida (http://joseafga.com.br)
	 * Creative Commons Attribution 3.0 Unported License (http://creativecommons.org/licenses/by/3.0/)
	 *
	 * Description:
	 *  Simple debugger for log triggered events from Modulider
	 *
	 * Syntax:
	 *  $('selector').modulider([options]).modDebug([fn]);
	 *  Function fn, function called to display message
	 */
	$.fn.modDebug = function (fn) {
		fn = fn || (window.console ? console.log : false) || $.noop;

		// fix webkit console
		var log = $.proxy(function () {
			var args = Array.prototype.slice.call(arguments);
			try {
				fn.apply(this, args);
			} catch (e) {
				fn.apply(console, args);
			}
		}, this);

		// create function for detach all debug events
		this.debugDestroy = function () {
			this.each(function () {
				$(this).off('.debug');
				log('Debug Destroyed', this);
			});
			return this;
		};

		// attach events
		return this.each(function () {
			log('Debug Initialized', this);
			$(this).on({
				'onPlay.debug onPause.debug onNext.debug onPrev.debug onFinishChanging.debug': function (e) {
					log(e.type);
				},
				'onStartChanging.debug': function (e, current, next) {
					log(e.type + ' - ' + current + ' to ' + next);
				}
			});
		});
	};

	/*!
	 * jQuery Modulider - Key Navigation plugin
	 *
	 * Version 0.1
	 * Copyright José Almeida (http://joseafga.com.br)
	 * Creative Commons Attribution 3.0 Unported License (http://creativecommons.org/licenses/by/3.0/)
	 *
	 * Description:
	 *  when press a key the corresponding function is performed
	 *  created for Modulider but may work in other ways
	 *
	 * Syntax:
	 *  $('selector').modulider([options]).modKeyNav([keys]);
	 *  Object keys, keyCode with string function name
	 */
	$.fn.modKeyNav = function (keys) {
		var slider = this;
		// default options
		keys = keys || {
			'37': 'prev', // <- go to previous slide
			'39': 'next' // -> go to next slide
		};

		$(window).keydown(function (e) {
			for (var k in keys) {
				if (e.keyCode == k) {
					e.preventDefault();
					slider[keys[k]]();
					break;
				}
			}
		});

		return this;
	};

	/*!
	 * jQuery Modulider - Caption Hide plugin
	 *
	 * Version 0.1
	 * Copyright José Almeida (http://joseafga.com.br)
	 * Creative Commons Attribution 3.0 Unported License (http://creativecommons.org/licenses/by/3.0/)
	 *
	 * Description:
	 *
	 * Syntax:
	 *  $('selector').modulider([options]);
	 */
	$.fn.modCaptionHide = function () {
		var slider = this;
		var isOver = false;

		slider.on('mouseenter',
			function () {
				isOver = true;
				slider.options.captions.eq(slider.currSlide).slideUp(500);
			}).on('onFinishChanging',
			function () {
				setTimeout(function () {
					if (isOver) slider.trigger('mouseenter');
				}, slider.options.speed);
			}).on('mouseleave',
			function () {
				isOver = false;
				slider.options.captions.eq(slider.currSlide).slideDown(500);
			}).on('onStartChanging', function () {
				slider.options.captions.eq(slider.currSlide).slideDown(500);
			});

		return this;
	};
})(jQuery);

src/jquery.modulider-latest.js
/*!
 * jQuery Modulider - Modular Slider
 *
 * Version 0.11.0
 * Copyright José Almeida (http://joseafga.com.br)
 * Creative Commons Attribution 3.0 Unported License (http://creativecommons.org/licenses/by/3.0/)
 * 
 * Description: 
 *  Modular, Easy and Flexible Slider for jQuery
 *  
 * Syntax:
 *  $(selector).modulider([options]);
 *  Object options
 * 
 * Depends:
 *	jQuery v1.7 or higher, http://jquery.com/
 */
(function($){
	/** =================== *
	 * CLASS
	 * ==================== */
	function Modulider($slider, opts){
		var o = $.extend({
			slides: $slider.children().first().children(),
			captions: $(), // captions element must be within slides and with class "caption"
			nav: $slider.children().last() // if you don't want thumbnails put false
		}, $.fn.modulider.defaults, opts || {});

		var ms = $.extend({}, $slider, {
			currSlide: 0,
			nextSlide: 0,
			running: false,
			inTransition: false,
			_ap: false, // overall autoplay: set true within the module to does not run default autoplay

			// slider options
			options: o,

			_init: function(){
				// distributes captions and set display for slides
				for (var i = 0, len = o.slides.length; i < len; i++) {
					var slide = o.slides.eq(i);
					if (i !== this.currSlide)
						slide.css('display', 'none');
					o.captions.push(slide.children('.caption')[0]);
				}

				// call all functions from options
				this.implement(o.modules || {});

				// autoplay option
				if (o.autoplay && !this._ap)
					this.play();

				return this;
			},

			_startTimer: function(){
				if (this.running) {
					this._clearTimer(); // prevent start multiples timing events
					// start interval
					this.timer = setTimeout($.proxy(function(){
						if (!this.inTransition)
							this.goTo(this.currSlide + 1);
					}, this), o.delay);
				}
				return this;
			},

			// clears current timer
			_clearTimer: function(){
				clearTimeout(this.timer);
				return this;
			},

			// modularity
			implement: obj2param(function(mod, value){
				var mods = $.fn.modulider.modules; // all modules
				if ($.isFunction(value))
					value = value.call(mods, index, mod);

				if (mod in mods && value !== false)
					mods[mod].call(this, typeof value === 'object' ? value || {} : {});

				return this;
			}),

			changeSlide: function(){
				this.trigger('onStartChanging', [this.nextSlide, this.currSlide]);
				// get next/prev slide and set z-index
				var next = o.slides.eq(this.nextSlide).css('zIndex', 1),
					prev = o.slides.eq(this.currSlide).css('zIndex', 0);
				// set caption inside next/prev
				next.caption = o.captions.eq(this.nextSlide).css('display', 'none');
				prev.caption = o.captions.eq(this.currSlide);
				this.currSlide = this.nextSlide;
				this.inTransition = true;

				$.fn.modulider.transitions[o.transition](o, next, prev); // start transition

				// callback logic inspired on jQuery UI Effects Explode
				// run approximately when the slide transition have finished
				setTimeout($.proxy(function(){
					prev.css('display', 'none');
					this.inTransition = false;
					this.trigger('onFinishChanging');
					if (this.currSlide != this.nextSlide)
						this._startTimer().changeSlide(); else
						this._startTimer();
				}, this), o.speed);
				return this;
			},

			play: function(){
				this.trigger('onPlay');
				this.running = true;
				return this._startTimer();
			},

			pause: function(){
				this.trigger('onPause');
				this.running = false;
				return this._clearTimer();
			},

			// show arbitrary slide
			goTo: function(to){
				// fixes the number to the range of slides
				to = (to >= o.slides.length) ? 0 : (to < 0) ? o.slides.length - 1 : to;
				if (this.currSlide != to) {
					this.nextSlide = to;
					if (!this.inTransition)
						this.changeSlide(); // change slide if transition is not in progress
				}
				return this;
			},

			// show next slide
			next: function(){
				this.trigger('onNext');
				return this.goTo(this.currSlide + 1);
			},

			// show previous slide
			prev: function(){
				this.trigger('onPrev');
				return this.goTo(this.currSlide - 1);
			}
		});

		return ms._init();
	}

	/** =================== *
	 * JQUERY PLUGIN
	 * ==================== */
	$.fn.modulider = function(opts){
		var sliders = null;

		this.each(function(index){
			var $slider = $(this),
				ms;

			// if element have an instance of plugin
			if ($slider.data('modulider')) {
				// retrieves the previously stored data
				ms = $slider.data('modulider');
			} else {
				ms = new Modulider($slider, opts);
				// initialize slider and store data
				$slider.data('modulider', ms);
			}

			if (sliders == null)
				sliders = ms; // make first slider functional
			else
				sliders.push(this); // push element only of the others
		});

		return sliders;
	};

	/** =================== *
	 * DEFAULTS
	 * ==================== */
	$.fn.modulider.defaults = {
		delay: 5000, // time for changing slides (milliseconds)
		speed: 500, // transition speed (milliseconds)
		transition: 'fade', // transition ($.fn.modulider.transitions)
		transitionCaption: 'none', // ???
		easing: 'linear',
		randomList: [], // list of possible transitions (used for random transition)
		autoplay: true // if false starts paused
		//, modules: {}
	};

	/** =================== *
	 * TRANSITIONS
	 * 9 transitions based on built-in jQuery effects + "random" and "none"
	 *
	 * if you will animate the caption must hide it before
	 * previous slide hide automatically in options.speed (milliseconds)
	 * ==================== */
	$.fn.modulider.transitions = obj2param(function(name, fn, add){
		var _trans = {};

		add || add === undefined && $.fn.modulider.defaults.randomList.push(name); // add to random list
		$.fn.modulider.transitions[name] = fn;
	});

	$.fn.modulider.transitions({
		random: function(opts, next, prev){
			// get a random value from options.randomList array
			this[opts.randomList[Math.floor(Math.random() * opts.randomList.length)]](opts, next, prev);
		},
		none: function(speed, next){
			next.css('display', '').caption.css('display', '');
		}
	}, false);

	// add multiple transitions
	var trans = {
		fade: ['fadeIn', 'fadeOut'],
		slide: ['slideDown', 'slideUp'],
		show: ['show', 'hide']
	};

	$.each(trans, function(name, fix){
		$.fn.modulider.transitions(name, function(opts, next){
			next.caption.css('visibility', 'hidden'); // ie6 fix

			next[fix[0]](opts.speed, opts.easing, function(){
				next.caption.css('visibility', 'visible'); // ie6 fix
				next.caption[trans[opts.transitionCaption] ? trans[opts.transitionCaption][0] : fix[0]](opts.speed / 2, opts.easing); // caption show in half time
				//$.fn.modulider.transitions._caption.show.call(next.caption);
			});
		});

		$.fn.modulider.transitions('cross' + name, function(opts, next, prev){
			prev[fix[1]](opts.speed, opts.easing);
			this[name](opts, next);
		});

		$.fn.modulider.transitions(name + 'background', function(opts, next, prev){
			var o = $.extend({}, opts, {
				speed: opts.speed / 2
			}); // copy options
			prev[fix[1]](o.speed, o.easing, $.proxy(function(){
				this[name](o, next);
			}, this));
		});
	});

	/** =================== *
	 * MODULES
	 * ==================== */
	$.fn.modulider.modules = {
		/**
		 * NAVIGATION
		 *
		 * options:
		 *   Selector nav
		 *   Boolean autoNav
		 */
		navigation: function(opts){
			opts.nav = this.find(opts.nav || '.navigation');

			// auto generete thumbs
			if (opts.autoNav) {
				// add thumbnails
				var thumbs = '';
				this.options.slides.each(function(){
					thumbs += ('<div class="thumb"><img src="' + $(this).find('img').attr('src') + '" alt="thumb" /></div>');
				});
				opts.nav.append(thumbs);
			}

			var childs = opts.nav.children();
			// add class 'current' to item nav of current slide
			childs.eq(this.currSlide).addClass('current');
			this.on('onStartChanging.modulider', function(e, next, prev){
				childs.eq(prev).removeClass('current');
				childs.eq(next).addClass('current');
			});
			// function when click on nav item
			childs.on('click.modulider', this, function(e){
				e.data.goTo($(this).index());
			});
		},

		/**
		 * MOUSE ENTER/LEAVE
		 */
		slideEvents: function(){
			this.options.slides.on('mouseenter.modulider', $.proxy(this.pause, this)).on('mouseleave.mSlider', $.proxy(this.play, this));
		},

		/**
		 * PRELOAD
		 *
		 * options:
		 *   String(HTML) template
		 */
		preload: function(template){
			template = template || '<div class="preload"><div class="progress"><div class="progress-bar"></div></div></div>';
			this._ap = true; // does not run default autoplay

			var $this = this,
				$load = $(template).appendTo(this),
				$bar = $load.find('.progress-bar'),
				$imgs = this.find('img'), imgLen = $imgs.length, imgLoaded = 0;

			$imgs.each(function(){
				this.complete ? updateProgress() : this.onload = this.onerror = updateProgress;
			});

			function updateProgress(){
				imgLoaded++;
				$bar.css('width', Math.round(imgLoaded * 100 / imgLen) + '%');

				if (imgLoaded == imgLen)
					$load.fadeOut($this.options.speed, function(){
						if ($this.options.autoplay)
							$this.play(); // autoplay option
					});
			}
		}
	};

	/** =================== *
	 * PRIVATE - OBJECT TO PARAMETER
	 * checks if parameter is a object and distribute in function key:value with another parameters
	 * ==================== */
	function obj2param(fn){
		return function(obj){
			if (typeof obj === 'object') {
				var args = Array.prototype.slice.call(arguments, 1);
				for (var k in obj) {
					fn.apply(this, $.merge([k, obj[k]], args));
				}
			} else
				fn.apply(this, arguments);
		}
	}
})(jQuery);